### DISCUSSION TITLE
Secp256r1

### DISCUSSION THREAD

**Discussion by SherLzp** [Consensys/gnark-crypto#367] [Q&A] [answered]
Hi sir @gbotrel @yelhousni , recently I want to add support for secp256r1 curve for gnark-crypto.

I'm fully understanding how the following value comes, could you please explain it, appreciate it:
1.  What's the `q` and `r` here?
https://github.com/ConsenSys/gnark-crypto/blob/develop/ecc/secp256k1/fp/element.go#L82

2. How about the values here:
https://github.com/ConsenSys/gnark-crypto/blob/develop/ecc/secp256k1/fp/element.go#L298-L301

And is there any other parts should I take care?

**Comment by yelhousni** [comment, marked as answer]
Hi @SherLzp !
1. `q` is the modulus `115792089237316195423570985008687907853269984665640564039457584007908834671663`and `r` is `2^64`. So `-1/q % 2^64` is `15580212934572586289`.
2. These are the 64-bit limbs of `(q-1)/2` in decimal. e.g. write `(q-1)/2` in binary and take the first 64 bits. This is `1111111111111111111111111111111111111111111111111111111111111111` in binary and `18446744073709551615` in decimal.

You shouldn't care about these as they are auto-generated. You just need to put the curve moduli in the config file (e.g. for [secp256k1](https://github.com/ConsenSys/gnark-crypto/blob/master/internal/generator/config/secp256k1.go)) and the `fp/` and `fr/` packages are auto-generated.

**Reply by SherLzp** [reply]
thanks!
